### ISSUE TITLE
Setting up CI on this repo

### ISSUE THREAD

**Issue by jakirkham** [conda/conda-recipes#441]
This came up in a separate discussion and I think it is worth pursuing. Namely, adding CI for this repo using some combination of AppVeyor and probably Travis.

attn: @patricksnape

**Comment by msarahan** [comment]
This is a great idea, but wow, what scope!  Every single package?  It seems to me that some sort of more clever deal that does CI jobs for each recipe independently is a better approach.

**Comment by jakirkham** [comment]
Agreed, it is quite large. I figure some discussion is required before any implementation is tried.

**Comment by jakirkham** [comment]
From what I have seen, the normal strategy is to detect only recipes that have changed and build only those ones and skip the rest. This is one common such thing that I have seen ( https://github.com/pelson/Obvious-CI ). There are other variants.

**Comment by patricksnape** [comment]
@jakirkham thanks for setting this up! Apologies for my slow responses - I'm working towards a very tight deadline at the moment so I'm going to be a bit unresponsive for a couple of weeks unfortunately. 

This sounds awesome. I'm not sure what the best way to handle incremental builds is... especially since Appveyor has a build queue of 1-job at a time, which means that every PR and every build target under a particular user gets queued in submission order. So if this repository became very effective (Which would be great), the build times good start spiralling out of control.

**Comment by jf---** [comment]
Having CI on the conda recipe repo is highly attractive indeed....

What if one accepts that the current master branch is stable? This assumption does imply that partial build are acceptable.

Now, when a commit triggers a build, it should be straightfwd to see what has altered in the commit.
We can think of conda as a protocol, and execute the protocol for the conda recipe's that were altered.
That is to say: can I run "conda build the-recipe-altered" succesfully?

Imagine that a recipe where CI hasnt ran yet fails. That still yields the functionality one seeks from a CI.

So coverage would complete over time.

I think that if splitting can be reasonably avoided, I think that's a forté, since it promotes interoperability between recipes. Since these are in constant flux, splitting repo's might invoke communication between the different recipe repo's. 

Part of the attraction of a CI is the implicit communication that stems from it, hence a large set of recipes [might be coherently managed this way](https://www.youtube.com/watch?v=W71BTkUbdqE).

Finally, [binstar has CI functionality](https://github.com/Anaconda-Server/anaconda-build), I believe @srossross is leading that development. The obvious question, would be where the great tradition of dogfeeding would fit in here

**Comment by jakirkham** [comment]
@patricksnape 

Of course. No worries.

Looking at how frequently PRs are sent to this repo, the build lengths of the more typical packages here, and the fact that there are already so many packages, I don't think this is so significant problem. Though this could change, I would be of the opinion that we wait and see.

Another consideration is that we don't use Appveyor or Travis for that matter and instead have this setup with Binstar build (or Anaconda build?). This would allow better control of the infrastructure and would be able to provide the diversity of required platforms. Additionally, this could allow for the option of uploading build artifacts that could be downloaded from some channel thus significantly cutting down on the maintenance burden (if this isn't already happening somehow behind the scenes) and allow for some added manual verification of packages that one is trying to build. Just to be clear, this would require Continuum's blessing.

As far as splitting goes, some are split out like so ( https://github.com/conda/conda-javascript-recipes ). I could see some value in doing more by language splits. For instances, there are a ton of R recipes here. It might be worth considering splitting things like Perl. When it comes to Python, I would like to see pure Python recipes split out in a repo that should ultimately be shrinking as I think `conda pipbuild` should eventually replace the need for this.

**Comment by jakirkham** [comment]
@jf--- 

Agreed.

> What if one accepts that the current master branch is stable? This assumption does imply that partial build are acceptable.
> 
> Now, when a commit triggers a build, it should be straightfwd to see what has altered in the commit.
We can think of conda as a protocol, and execute the protocol for the conda recipe's that were altered.
That is to say: can I run "conda build the-recipe-altered" succesfully?

That's what I am proposing. Though I didn't state it as directly as you have. ( https://github.com/conda/conda-recipes/issues/441#issuecomment-149623053 ). The protocol I would recommend is the one cited previously it is already actively used and relatively stable.

> Imagine that a recipe where CI hasnt ran yet fails. That still yields the functionality one seeks from a CI. So coverage would complete over time.

Agreed.

> I think that if splitting can be reasonably avoided, I think that's a forté, since it promotes interoperability between recipes. Since these are in constant flux, splitting repo's might invoke communication between the different recipe repo's.

That is a fair point. However, if split along language lines, this might not be as much of an issue. It's also possible to take on a `git submodule` style solution here.

> Finally, binstar has CI functionality, I believe @srossross is leading that development. The obvious question, would be where the great tradition of dogfeeding would fit in here 

I also like this idea and would provide some more options for us in terms of getting things working.

**Comment by jf---** [comment]
Hi @jakirkham 

> split along language lines

This might not be as intuitive as the proposition seems. For instance, there's a ton of code on conda that wraps C / C++ / Fortran libs. Python has created a polyglot world, where interoperability between languages is increasingly freed from friction. That said, a separate R or javascript repo might make sense. 

> I also like this idea and would provide some more options for us in terms of getting things working and > playing with or using builds. This would require Continuum's blessing, of course.

Perhaps @ccordoba12 or @teoliphant are willing to add to the discussion?

**Comment by ccordoba12** [comment]
I've been following this discussion with great interest, but setting CI integration for all recipes here seems a daunting task. Perhaps we could try to support the `conda-forge` project

https://github.com/conda-forge

which was set up to solve this problem.

----

Pinging @pelson about it.

**Comment by jakirkham** [comment]
I have also heard of `conda-forge`. This may be the better way in the end. It would be nice to know what the relationship will be between Continuum and `conda-forge` (if there will be one).

**Comment by ccordoba12** [comment]
There is no relation at the moment, i.e. conda-forge is an independent 
project :-)

**Comment by jakirkham** [comment]
Right. So, the reason I ask is this repo has some authority because it is maintained by Continuum (unless I am badly mistaken). However, if we are thinking that CI is not feasible here and that it may be better approached on `conda-forge`.

Should it then be suggested or recommended that community recipes get contributed to `conda-forge`? Should there be some sort of migration recipes from here over to `conda-forge`, which raises this question ( https://github.com/conda/conda-recipes/issues/37#issuecomment-148558373 )? What will the long term plan of this repo be in relation to `conda-forge`?

**Comment by pelson** [comment]
> if we are thinking that CI is not feasible

CI is feasible - there are a number of places where this is working pretty successfully: 

 * https://github.com/ioos/conda-recipes
 * https://github.com/scitools/conda-recipes-scitools
 * https://github.com/conda-forge/staged-recipes

They are all using the same technology under the hood (https://github.com/pelson/Obvious-CI). We check to see if a build exists already on the appropriate anaconda.org channel, and if it doesn't we build it. If the commit is a merge we also upload to anaconda.org. This works really well - we've gone through ~1 year of iteration on this, and have solved most of the big challenges. The one remaining sticky issue for us is that you will notice there is a degree of overlap in the recipes which exist in those repos - we haven't figured out a *good* strategy for managing that. The one solution that has some really pleasant generalisations is to fall back to the concept of one recipe, one repository - that is what ``conda-forge`` does. Of course, that is a *crazy* idea as you have to manage so many repos - that is what ``conda-smithy`` is designed to alleviate. It's all pretty organic at the moment, and not really fit for a large-scale migration, but the more hands at the pump the merrier (https://github.com/conda-forge/conda-forge.github.io/issues).

Pinging @ocefpaf as he is doing an incredible job of curating the IOOS channel.

**Comment by ocefpaf** [comment]
Setting CIs for the entire repo is doable, as @pelson said, and [Obvious-CI](https://github.com/pelson/Obvious-CI) is the best tool for the job.

Here are a few comments, based on our experience, that address some of the points people raised above:

- The disadvantage of slow CIs, like AppVeyor, is not a big issue after the first round of builds are uploaded.  Obvious-CI takes care of building only the new recipes, build number, or new version. Although it **does** makes sense to split the repository in themes like conda-recipes-{JS,R,etc}.  Note that it is OK to let the PR sit for a while before merging.  In my opinion AppVeyor is usually faster than our human reviewers :wink:

- The [scitools](https://github.com/scitools/conda-recipes-scitools) and [ioos](https://github.com/ioos/conda-recipes) uses Appveyor for Windows, Travis-CI for OSX, and CircleCI for Linux.  We have more control with CircleCI because that uses a [Dockerfile](https://github.com/pelson/Obvious-CI/blob/master/obvious-ci.docker/linux64_obvci/Dockerfile).  But again control is not a problem. It tuns out that AppVeyor and Travis-CI are robust enough when it comes to providing the necessary build tools.  They are also field tested when it comes to creating binaries that are installable in a variate of machines configurations. I  taught a Python course where people had WinXP, Win7-10, OSX, and many flavors of Linux.  Everybody installed the necessary requirements from  the ioos channel without a problem using a single `environment.yml` file.

However, keeping such CI configuration is a lot of work! There are many moving parts that are constantly breaking!!  It is almost a full time job. And there is the recipe overlap.

I guess that people will be more prone to contribute to one or a small subset of recipes. A huge repository with tons of recipes can be intimidating even if someone is just updating a single software version.

In the long run I believe that having 1-repo 1-recipe, under a GitHub organization like [conda-forge](https://github.com/conda-forge/conda-forge.github.io), is more attractive to contributors.

PS: What would be even better is if conda-forge becomes a place for recipes that will be incorporated by the default channel in the long run.

**Comment by jakirkham** [comment]
So, based on what you guys have said it sounds like `conda-forge` is pretty cool and it would be nice for interested people to help out. However, we aren't at the stage where this magnitude of recipes should find their way into `conda-forge`.

At present, I remain in favor of getting CI going here using `Obvious-CI`. Though I think there are some outstanding questions about licensing and how that affects what is being done here.

>  Note that it is OK to let the PR sit for a while before merging. In my opinion AppVeyor is usually faster than our human reviewers :wink:

I strongly agree with this and would add the biggest problem now is having simple reproducible environments for builds, which this provides. 

The use of CentOS for Linux is reassuring as this tends to be a nice lowest common denominator. Of course, there are always corner cases. Feel free to look at the work done on `gcc` ( https://github.com/conda/conda-recipes/pull/279 ) if you want to see what I mean. It is probably worth discussing whether it should still be CentOS 6 or whether we can switch to say CentOS 5.11 to be a bit more compatible with existing packages built here.

Knowing that AppVeyor doesn't hurt so much on a group of recipes like `ioos` is very reassuring.

Though there were a lot of good points made regarding simplicity with a repo per recipe and the like, I think these sorts of rearrangements are secondary to getting CI working. Once CI is working it is much easier to verify that these sorts of splits are made correctly.

This raises another question. Does `Obvious-CI` check the dependency graph to make sure it is ok even if it is not building all the recipes?

**Comment by jf---** [comment]
A thought that comes to mind is whether it would be possible to invoke a CI build, when a dependency has changed. Imagine my image processing libs relies on numpy. Now, if a change to numpy would trigger a build of my image processing lib, you'd be informed the moment incompatibilities arise.

**Comment by ocefpaf** [comment]
> However, we aren't at the stage where this magnitude of recipes should find their way into `conda-forge`.

Agreed.  But a select few, that are essential to a larger community, would be most welcome.  Specially those that are currently broken in the default channel.  (Like `pyproj` for windows.)

>  It is probably worth discussing whether it should still be CentOS 6 or whether we can switch to say CentOS 5.11 to be a bit more compatible with existing packages built here.

IOOS and Scitools packages are built against CentOS 6. I have some people reporting that the packages worked in CentOS 5.  I never tested that myself though.

> Knowing that AppVeyor doesn't hurt so much on a group of recipes like ioos is very reassuring.

AppVeyor is actually quite surprising.  They have cygwin, MSYS, MinGW, and even miniconda pre-installed!

> I think these sorts of rearrangements are secondary to getting CI working. 

I disagree with that.  Setting the CIs for all recipes here will be a different task than setting CIs using `conda-smithy`.  Here is an example of what I mean:

I had a package `X` with a build dependency on `gsl`, so I pulled `gsl` from Travis-CI "repo" and the package built just fine.  I second package `Y` had a run dependency on `gsl`.  Because the CI was pulling `gsl` already, package `Y` built and passed tests.  However, when installing this second package with conda on a computer without `gsl` it crashed.

The 1-repo to 1-recipe model would have avoided that "CI contamination."  At least it would have made the debugging process easier.

This does not mean that I am against implementing the obvious-ci workflow here, but I am in the process of "dismantling" the IOOS channel in favor of conda-forge.  The future I envision for the IOOS channel is to build only a few domain specific packages while the bulk of the binaries would be copied from `conda-forge`.

> Does Obvious-CI check the dependency graph to make sure it is ok even if it is not building all the recipes?

At the first run Obvious-CI will build everything.  Get ready for a lot of CI restarting until the first pass of builds happen!  After that it will build only the modified ones.

> A thought that comes to mind is whether it would be possible to invoke a CI build, when a dependency has changed.

I have thought a lot about that.  I was used to a Linux build system that did just that and it was awesome.  However, implementing that for conda+obvious-ci+(Travis-CI, AppVeyor, CircleCI) might be too much.  For example: a numpy update would be similar to a "first build load" on the CIs, requiring a lot of manual re-starts.  And some of the builds will be unnecessary, since only packages that depends on numpy ABI or those that require a new feature, actually need to be re-build.

**Comment by patricksnape** [comment]
This all seems like really great info! Having just caught myself up, I think I agree with @ocefpaf in that I believe the separate repositories may be easier to maintain in the long run. Particularly if certain projects can maintain 'champions' that ensure that those projects are consistently building and up to date. The major disadvantage of the monolithic repository is that if something requires a very specific change to the CI system in order to build then it may block builds across other recipes. Also, as @ocefpaf and @jf--- mentioned, the initial overhead of getting the repository set up may be incredibly high. `conda-forge` looks really really interesting - is there anything we can do that might make testing the integration between conda-recipes and conda-forge easier?
